vllm/model_executor/layers/layernorm.py

```diff
@@ -185,47 +180,69 @@ def __init__(
                 with_fused_add=True, dtype=weight_dtype
             )
 
-    def forward_native(
-        self,
+    @staticmethod
+    def forward_static(
         x: torch.Tensor,
+        variance_epsilon: float,
+        hidden_size: int,
+        orig_dtype: torch.dtype,
+        weight: torch.Tensor | None = None,
         residual: torch.Tensor | None = None,
+        variance_size_override: int | None = None,
     ) -> torch.Tensor | tuple[torch.Tensor, torch.Tensor]:
         """PyTorch-native implementation equivalent to forward()."""
-        orig_dtype = x.dtype
         x = x.to(torch.float32)
         if residual is not None:
-            x = x + residual.to(torch.float32)
+            # residual promoted f16->f32 automatically,
+            # otherwise Inductor eliminates the casts to and from f16,
+            # increasing memory usage (and complicating pattern matching)
+            x = x + residual
```

Comment:
NOTE: this change in RMSNorm helps matching, and also reduces memory footprint cc @ilmarkov

Reply:
Discussed offline: the change is from 1 -> 2:
1. `x.to(float32) + residual.to(float32)`
1. `x.to(float32) + residual`

In the second case (this PR), residual should get promoted automatically.